docs/rfcs/xxx-pretranspile.md

```diff
@@ -0,0 +1,44 @@
+# Pre-transpiled Atom packages
+
+## Status
+
+Proposed
+
+## Summary
+
+This feature will enable package authors to use conventional npm tooling and package.json conventions to take advantage of JavaScript transpilers like Babel or TypeScript.
+
+## Motivation
+
+Transpiling packages on _publish_ rather than _load_ will have great benefits for package authors:
+
+* Standard `npm` tooling like `prepare` scripts will work for apm packages exactly as they work for npm packages. This will remove the need for custom transpiler pipeline modules like [atom-babel6-transpiler](https://github.com/atom/atom-babel6-transpiler) or [atom-typescript-transpiler](https://github.com/smhxx/atom-ts-transpiler) with their own, independent documentation, configuration and setup.
+* Packages can move transpiler-related dependencies to `devDependencies` and trim installation bloat substantially. (as a data point, the TypeScript compiler is 30MB.)
+* First-time package load will no longer take a hit from transpiling all of the source into the cache. All package loads will benefit from no longer needing to check and side-load from the compiler cache.
```

Comment:
Is this actually true? I thought that Atom currently transpiled _everything_ down to a very basic ES5 in the compile-cache, the custom transpilation feature just lets you control that. If you're talking about this RFC bringing Atom to running the package code directly as it is shipped that could cause some rather subtle issues with packages as we have seen in the past they sometimes rely (accidentally or on purpose) on quirks of this "always transpiled" path.

Reply:
> I thought that Atom currently transpiled everything down to a very basic ES5 in the compile-cache, the custom transpilation feature just lets you control that.

Nope! JavaScript files have to explicitly opt in to the built-in Babel transpiler with a `/** @babel */` comment.

https://github.com/atom/atom/blob/e500f933e05e53286527846f008dac78f0c83b13/src/babel.js#L22-L27

When `shouldCompile()` returns `false`, the source is passed through and required as-is:

https://github.com/atom/atom/blob/e500f933e05e53286527846f008dac78f0c83b13/src/compile-cache.js#L84-L99

> If you're talking about this RFC bringing Atom to running the package code directly as it is shipped that could cause some rather subtle issues with packages as we have seen in the past they sometimes rely (accidentally or on purpose) on quirks of this "always transpiled" path.

Eventually we _might_ be able to eliminate the compile cache entirely, at least in non-dev mode, but we couldn't think about that until enough packages transition to using `prepare` steps and re-publish as tarballs anyway.

We'll have to leave it in for the foreseeable future to handle old packages, though, so we won't be able to get the performance boost of not having to redundantly slurp the source code of every module when it's required. I'll take that second sentence out, then.